kraken: trying to reduce build time and memory

kinnou02 opened the pull request:
Building data.cpp takes a very long time (more than 2minute) and a lot
of memory (around 4.5go). This seems to be caused by boost serialize.

In this PR I have started to move serialization function in .cpp to
limit the need of rebuilding it every time. I also moved each type in
its own unit to reduce memory usage and prevent the rebuild of
everything. It should also speed up the build as each unit will be
smaller and build concurrently.

Also, I have added some forward-declare in `data.h`, so a lot of files had to be updated.

kinnou02 commented:
I forgot to put it in draft :cry:

xlqian commented on source/type/indexes.h:
```diff
@@ -0,0 +1,25 @@
+#pragma once
```
copy right?

xlqian commented on source/type/stop_point.cpp:
```diff
@@ -0,0 +1,45 @@
+
```
copy right

nonifier commented on source/type/data.h:
```diff
@@ -29,47 +29,28 @@ www.navitia.io
 #include <boost/utility.hpp>
+#include <boost/serialization/serialization.hpp>
+#include <boost/serialization/shared_ptr.hpp>
+//#include <boost/serialization/export.hpp>
```
this probably needs to be removed ;-)

nonifier commented on source/type/fdw_type.h:
```diff
@@ -0,0 +1,84 @@
+/* Copyright © 2001-2014, Canal TP and/or its affiliates. All rights reserved.
```
Bouyaaaaaaaa :fireworks: !!!

nonifier reviewed (approved):
:+1:

kinnou02 committed:
kraken: trying to reduce build time and memory

building data.cpp is take a very long time (more than 2minute) and a lot
of memory (around 4.5go). This seems to be caused by boost serialize.

In this PR I have started to move serialization function in .cpp to
limit the need of rebuilding it every time. I also moved each type in
it's own unit to reduce memory usage and prevent the rebuild of
everything. It should also speed up the build as each unit will be
smaller and build concurrently.

Also, I have added some forward declare in data.h, so a lot of files had to be updated

pbougue commented on source/type/fdw_type.h:
```diff
@@ -0,0 +1,81 @@
+struct JourneyPatternPoint;
+}  // namespace routing
+namespace type {
+class MetaData;
```
Clang warning: please forward it as a `struct` to match the one from timezone_manager.h
```suggestion
struct MetaData;
```

pbougue commented on source/type/data.h:
```diff
@@ -29,19 +29,20 @@ www.navitia.io
 #include "utils/obj_factory.h"
 #include "utils/ptime.h"
-#include "georef/adminref.h"
+#include "type/fdw_type.h"
```
File name should be f`wd`_type.h no?

kinnou02 replied:
it should :cry:

pbougue commented on source/type/data.cpp:
```diff
@@ -389,94 +389,6 @@ void Data::complete() {
-    return validity;
-}
-
-void Data::build_associated_calendar() {
```
Shouldn't you remove the commented line 361 above, then?
Looks like it was ~moved~ copied into ed in this PR: https://github.com/CanalTP/navitia/pull/831/files#r23221968 :wink:

pbougue reviewed (approved):
Didn't re-review the whole commit :)

kinnou02 committed:
Move calendar in its own file

kinnou02 committed:
Update fwd_type.h
